code/lib/builder-vite/src/plugins/externals-plugin.ts

```diff
@@ -0,0 +1,112 @@
+import { join } from 'node:path';
+import { init, parse } from 'es-module-lexer';
+import MagicString from 'magic-string';
+import { emptyDir, ensureDir, ensureFile, writeFile } from 'fs-extra';
+import { mergeAlias } from 'vite';
+import type { Alias, Plugin } from 'vite';
+import { globals } from '@storybook/preview/globals';
+
+type SingleGlobalName = keyof typeof globals;
+type Globals = typeof globals & Record<string, string>;
+
+const replacementMap = new Map([
+  ['import ', 'const '],
+  ['import{', 'const {'],
+  [' as ', ': '],
+  [' from ', ' = '],
+  ['}from', '} ='],
+]);
+
+/**
+ * This plugin swaps out imports of pre-bundled storybook preview modules for destructures from global
+ * variables that are added in runtime.mjs.
+ *
+ * For instance:
+ *
+ * ```js
+ * import { useMemo as useMemo2, useEffect as useEffect2 } from "@storybook/preview-api";
+ * ```
+ *
+ * becomes
+ *
+ * ```js
+ * const { useMemo: useMemo2, useEffect: useEffect2 } = __STORYBOOK_MODULE_PREVIEW_API__;
+ * ```
+ *
+ * It is based on existing plugins like https://github.com/crcong/vite-plugin-externals
+ * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our simple needs.
+ */
+export async function externalsPlugin() {
```

Comment:
Personally, even though this is only used for this one case right now, globals should probably still be an argument for this plugin instead of being imported directly from the `@storybook/preview/globals` package. More flexible for reuse in the case that the args need to be changed or this plugin needs to be used for another set of globals.
```suggestion
export async function externalsPlugin(externals: Record<string, string>) {
```

Reply:
Yes, I agree, that is just nice separation of concerns!